File: web/src/types/cluster.ts
// Cluster data types matching backend models

export interface ClusterInfo {
  cluster_id: number
  algorithm: string
  label: string
  size: number
  density: number
  avg_centrality: number
  representative_id: string
  representative_source_type: 'doc' | 'git'
  representative_title: string
  representative_path: string
  representative_classification: number
}

export interface ArticleClusterAssignment {
  cluster_id: number
  cluster_label: string
  centrality_score: number
  algorithm: string
  calculated_at: string
}

export interface GetClustersResponse {
  clusters: ClusterInfo[]
  total: number
  algorithm: string
}

export interface RunClusteringRequest {
  algorithm: string
}

export interface RunClusteringResponse {
  success: boolean
  message: string
  cluster_count: number
  algorithm: string
}

File: backend/internal/models/cluster.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// ClusterInfo represents summary information about a cluster
type ClusterInfo struct {
	ClusterID                     int               `json:"cluster_id"`
	Algorithm                     string            `json:"algorithm"`
	Label                         string            `json:"label"`
	Size                          int               `json:"size"`
	Density                       float64           `json:"density"`
	AvgCentrality                 float64           `json:"avg_centrality"`
	RepresentativeID              uuid.UUID         `json:"representative_id"`
	RepresentativeSourceType      ArticleSourceType `json:"representative_source_type"`
	RepresentativeTitle           string            `json:"representative_title"`
	RepresentativePath            string            `json:"representative_path"`
	RepresentativeClassification  int               `json:"representative_classification"`
}

// ArticleClusterAssignment represents a single article's cluster assignment
type ArticleClusterAssignment struct {
	ClusterID       int       `json:"cluster_id"`
	ClusterLabel    string    `json:"cluster_label"`
	CentralityScore float64   `json:"centrality_score"`
	Algorithm       string    `json:"algorithm"`
	CalculatedAt    time.Time `json:"calculated_at"`
}

// GetClustersResponse represents the API response for clusters endpoint
type GetClustersResponse struct {
	Clusters  []ClusterInfo `json:"clusters"`
	Total     int           `json:"total"`
	Algorithm string        `json:"algorithm"`
}

// RunClusteringRequest represents request to run clustering algorithm
type RunClusteringRequest struct {
	Algorithm string `json:"algorithm"` // 'label_propagation', etc.
}

// RunClusteringResponse represents response from clustering operation
type RunClusteringResponse struct {
	Success      bool   `json:"success"`
	Message      string `json:"message"`
	ClusterCount int    `json:"cluster_count"`
	Algorithm    string `json:"algorithm"`
}
